robosuite/models/objects/interactive_objects.py

```diff
@@ -0,0 +1,255 @@
+            contacting_body_ids = EU.all_contacting_body_ids(self.sim, self.body_id)
+
+        self._on_contact(contacting_body_ids)
+        self._on_customized_condition()
+
+
+class StatefulObject(object):
+    """Base class for object that maintains state"""
+    def __init__(self, all_state_names, default_state):
+        self._default_state = deepcopy(default_state)
+        self._all_state_names = copy(all_state_names)
+        self._state = self.default_state
+
+        assert(isinstance(default_state, dict))
+        assert(sorted(self._default_state.keys()) == sorted(self._all_state_names))
+
+        self._on_enter_state_funcs = []  # called once when switching to the state
+        self._on_exit_state_funcs = []  # called once when leaving the state
+
+        self._on_state_funcs = []  # gets called when the object is in the specified state
+        self._on_not_state_funcs = []  # gets called when the object is NOT in the specified state
+
+        self._set_state_behaviors()
+
+        self.reset()
+
+    def _set_state_behaviors(self):
+        """Fills in state behaviors"""
+        raise NotImplementedError
+
+    def add_on_enter_state_funcs(self, cond, func):
+        assert(self.is_valid_condition(cond))
+        self._on_enter_state_funcs.append((deepcopy(cond), func))
+
+    def add_on_exit_state_funcs(self, cond, func):
+        assert(self.is_valid_condition(cond))
+        self._on_exit_state_funcs.append((deepcopy(cond), func))
+
+    def add_on_state_funcs(self, cond, func):
+        assert(self.is_valid_condition(cond))
+        self._on_state_funcs.append((deepcopy(cond), func))
+
+    def add_on_not_enter_state_funcs(self, cond, func):
+        assert(self.is_valid_condition(cond))
+        self._on_not_state_funcs.append((deepcopy(cond), func))
+
+    def reset(self):
+        # reset state without triggering state functions
+        self._state = self.default_state
+
+    def is_valid_state(self, state):
+        return sorted(list(state.keys())) == sorted(self._all_state_names)
+
+    def is_valid_condition(self, cond):
+        for k in cond:
+            if k not in self._all_state_names:
+                return False
+        return True
+
+    @property
+    def default_state(self):
+        return deepcopy(self._default_state)
+
+    @property
+    def state(self):
+        return deepcopy(self._state)
+
+    @state.setter
+    def state(self, new_state):
+        assert(self.is_valid_state(new_state))
+
+        prev_state = self.state
+        self._state = deepcopy(new_state)
+
+        for cond, func in self._on_enter_state_funcs:
+            if satisfies(self.state, cond) and not satisfies(prev_state, cond):
+                func()
+
+        for cond, func in self._on_exit_state_funcs:
+            if not satisfies(self.state, cond) and satisfies(prev_state, cond):
+                func()
+
+
+class StatefulInteractiveObject(StatefulObject, InteractiveObject):
```

Comment:
could you explain the need for multiple inheritance here?

Reply:
It doesn't have to be. Just thought this makes more sense because StatefulObject and InteractiveObject serve different purposes.